File: backend_guide/app/api/v1/recipes.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status

try:
    from supabase import Client
except Exception:

    class Client:  # type: ignore
        def __init__(self, *args, **kwargs):
            pass


from app.api.deps import get_user_client, get_service_client, get_current_user
from app.schemas.auth import CurrentUser
from app.schemas.recipe import Recipe, RecipeCreate, RecipeUpdate

router = APIRouter()


@router.get("", response_model=List[Recipe])
def list_recipes(
    limit: int = 10,
    offset: int = 0,
    cube_tags: Optional[str] = None,
    dietary_tags: Optional[str] = None,
    current_user: CurrentUser = Depends(get_current_user),
    user_client: Client = Depends(get_user_client),
):
    """
    List recipes with optional filtering and pagination.
    Uses the user-scoped client.
    """
    try:
        query = user_client.from_("recipes").select("*")

        if cube_tags:
            tags_list = [t.strip() for t in cube_tags.split(",") if t.strip()]
            if tags_list:
                query = query.contains("cube_tags", tags_list)

        if dietary_tags:
            tags_list = [t.strip() for t in dietary_tags.split(",") if t.strip()]
            if tags_list:
                query = query.contains("dietary_tags", tags_list)

        query = query.range(offset, offset + limit - 1)
        res = query.execute()
        return res.data
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to list recipes: {str(e)}",
        )


@router.get("/{id}", response_model=Recipe)
def get_recipe(
    id: str,
    current_user: CurrentUser = Depends(get_current_user),
    user_client: Client = Depends(get_user_client),
):
    """
    Get a single recipe by ID.
    Uses the user-scoped client.
    """
    try:
        res = user_client.from_("recipes").select("*").eq("id", id).execute()
        if not res.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Recipe not found or access denied.",
            )
        return res.data[0]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to fetch recipe: {str(e)}",
        )


@router.post("", response_model=Recipe, status_code=status.HTTP_201_CREATED)
def create_recipe(
    body: RecipeCreate,
    current_user: CurrentUser = Depends(get_current_user),
    service_client: Client = Depends(get_service_client),
):
    """
    Create a new recipe.
    Requires cms_editor role. Uses service-role client.
    """
    if current_user.role != "cms_editor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Requires cms_editor role.",
        )

    try:
        insert_data = body.model_dump()
        res = service_client.from_("recipes").insert(insert_data).execute()
        if not res.data:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Failed to create recipe.",
            )
        return res.data[0]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to create recipe: {str(e)}",
        )


@router.put("/{id}", response_model=Recipe)
def update_recipe(
    id: str,
    body: RecipeUpdate,
    current_user: CurrentUser = Depends(get_current_user),
    service_client: Client = Depends(get_service_client),
):
    """
    Update a recipe.
    Requires cms_editor role. Uses service-role client.
    """
    if current_user.role != "cms_editor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Requires cms_editor role.",
        )

    update_data = body.model_dump(exclude_unset=True)
    if not update_data:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No update fields provided.",
        )

    try:
        res = service_client.from_("recipes").update(update_data).eq("id", id).execute()
        if not res.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Recipe not found or update failed.",
            )
        return res.data[0]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to update recipe: {str(e)}",
        )


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_recipe(
    id: str,
    current_user: CurrentUser = Depends(get_current_user),
    service_client: Client = Depends(get_service_client),
):
    """
    Delete a recipe.
    Requires cms_editor role. Uses service-role client.
    """
    if current_user.role != "cms_editor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Requires cms_editor role.",
        )

    try:
        service_client.from_("recipes").delete().eq("id", id).execute()
        return None
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to delete recipe: {str(e)}",
        )

File: backend/supabase/functions/api/routes/recipes.ts
import { Hono } from 'https://deno.land/x/hono@v3.11.8/mod.ts'
import { getServiceClient, getUserClient, getCurrentUser } from '../deps.ts'

const router = new Hono()

router.get('/', async (c) => {
  const client = getServiceClient()
  const limit = parseInt(c.req.query('limit') || '10')
  const offset = parseInt(c.req.query('offset') || '0')

  try {
    const { data, error } = await client
      .from('recipes')
      .select('*, metrics:recipe_metrics(*)')
      .range(offset, offset + limit - 1)

    if (error) {
      return c.json({ detail: `Failed to list recipes: ${error.message}` }, 400)
    }

    const mapped = (data || []).map(r => {
      const metrics = (r as any).metrics
      return {
        id: r.id,
        author_id: r.author_id,
        title: r.title,
        slug: r.slug,
        description: r.description,
        status: r.status,
        published_at: r.published_at,
        prep_time_seconds: metrics?.prep_time_seconds || 0,
        cook_time_seconds: metrics?.cook_time_seconds || 0,
        servings_default: metrics?.servings || 1,
        servings: metrics?.servings || 1,
        difficulty: metrics?.difficulty_level === 'easy' ? 1 : metrics?.difficulty_level === 'medium' ? 2 : 3,
        difficulty_level: metrics?.difficulty_level || 'easy',
        cube_tags: [],
        dietary_tags: []
      }
    })

    return c.json(mapped)
  } catch (err) {
    return c.json({ detail: `Failed to list recipes: ${(err as any).message}` }, 400)
  }
})

router.get('/:id', async (c) => {
  const id = c.req.param('id')
  const client = getUserClient(c)

  const { data, error } = await client
    .from('recipes')
    .select('*, metrics:recipe_metrics(*)')
    .eq('id', id)
    .single()

  if (error) {
    return c.json({ detail: 'Recipe not found' }, 404)
  }

  const metrics = (data as any).metrics
  return c.json({
    id: data.id,
    author_id: data.author_id,
    title: data.title,
    slug: data.slug,
    description: data.description,
    status: data.status,
    published_at: data.published_at,
    prep_time_seconds: metrics?.prep_time_seconds || 0,
    cook_time_seconds: metrics?.cook_time_seconds || 0,
    servings_default: metrics?.servings || 1,
    servings: metrics?.servings || 1,
    difficulty: metrics?.difficulty_level === 'easy' ? 1 : metrics?.difficulty_level === 'medium' ? 2 : 3,
    difficulty_level: metrics?.difficulty_level || 'easy',
    cube_tags: [],
    dietary_tags: []
  })
})

router.post('/', async (c) => {
  const currentUser = await getCurrentUser(c)
  if (currentUser.role !== 'cms_editor') {
    return c.json({ detail: 'Forbidden: Requires cms_editor role.' }, 403)
  }

  const serviceClient = getServiceClient()
  const body = await c.req.json()

  const recipePayload = {
    title: body.title,
    slug: body.slug || body.title?.toLowerCase().replace(/ /g, '-'),
    description: body.description || '',
    status: body.status || 'draft',
    published_at: body.published_at || null,
    author_id: currentUser.id
  }

  try {
    const { data: recipeData, error: recipeErr } = await serviceClient
      .from('recipes')
      .insert(recipePayload)
      .select()

    if (recipeErr || !recipeData || recipeData.length === 0) {
      return c.json({ detail: `Failed to create recipe: ${recipeErr?.message || 'Database error'}` }, 400)
    }

    const recipe = recipeData[0]
    const difficultyLevel = body.difficulty === 1 ? 'easy' : body.difficulty === 2 ? 'medium' : body.difficulty_level || 'easy'
    const metricsPayload = {
      recipe_id: recipe.id,
      prep_time_seconds: body.prep_time_seconds || (body.prep_time_minutes ? body.prep_time_minutes * 60 : 0),
      cook_time_seconds: body.cook_time_seconds || (body.cook_time_minutes ? body.cook_time_minutes * 60 : 0),
      servings: body.servings || body.servings_default || 1,
      difficulty_level: difficultyLevel
    }

    await serviceClient.from('recipe_metrics').insert(metricsPayload)

    return c.json({
      id: recipe.id,
      ...recipe,
      ...metricsPayload,
      servings_default: metricsPayload.servings
    }, 201)
  } catch (err) {
    return c.json({ detail: `Failed to create recipe: ${(err as any).message}` }, 400)
  }
})

router.put('/:id', async (c) => {
  const id = c.req.param('id')
  const serviceClient = getServiceClient()
  const body = await c.req.json()
  if (!body || Object.keys(body).length === 0) {
    return c.json({ detail: 'No update fields provided.' }, 400)
  }

  const recipeUpdate: any = {}
  if (body.title !== undefined) recipeUpdate.title = body.title
  if (body.slug !== undefined) recipeUpdate.slug = body.slug
  if (body.description !== undefined) recipeUpdate.description = body.description
  if (body.status !== undefined) recipeUpdate.status = body.status
  if (body.published_at !== undefined) recipeUpdate.published_at = body.published_at

  const metricsUpdate: any = {}
  if (body.prep_time_seconds !== undefined) metricsUpdate.prep_time_seconds = body.prep_time_seconds
  if (body.cook_time_seconds !== undefined) metricsUpdate.cook_time_seconds = body.cook_time_seconds
  if (body.servings !== undefined) metricsUpdate.servings = body.servings
  if (body.servings_default !== undefined) metricsUpdate.servings = body.servings_default
  if (body.difficulty !== undefined) {
    metricsUpdate.difficulty_level = body.difficulty === 1 ? 'easy' : body.difficulty === 2 ? 'medium' : 'hard'
  }
  if (body.difficulty_level !== undefined) metricsUpdate.difficulty_level = body.difficulty_level

  try {
    let recipe = null
    if (Object.keys(recipeUpdate).length > 0) {
      const { data, error } = await serviceClient.from('recipes').update(recipeUpdate).eq('id', id).select()
      if (error || !data || data.length === 0) {
        return c.json({ detail: 'Recipe not found or update failed.' }, 404)
      }
      recipe = data[0]
    } else {
      const { data } = await serviceClient.from('recipes').select('*').eq('id', id)
      if (data && data.length > 0) recipe = data[0]
    }

    if (!recipe) {
      return c.json({ detail: 'Recipe not found.' }, 404)
    }

    if (Object.keys(metricsUpdate).length > 0) {
      await serviceClient.from('recipe_metrics').update(metricsUpdate).eq('recipe_id', id)
    }

    const { data: finalMetrics } = await serviceClient.from('recipe_metrics').select('*').eq('recipe_id', id)
    const metrics = finalMetrics && finalMetrics.length > 0 ? finalMetrics[0] : {}

    return c.json({
      id: recipe.id,
      author_id: recipe.author_id,
      title: recipe.title,
      slug: recipe.slug,
      description: recipe.description,
      status: recipe.status,
      published_at: recipe.published_at,
      prep_time_seconds: metrics.prep_time_seconds || 0,
      cook_time_seconds: metrics.cook_time_seconds || 0,
      servings: metrics.servings || 1,
      servings_default: metrics.servings || 1,
      difficulty_level: metrics.difficulty_level || 'easy'
    })
  } catch (err) {
    return c.json({ detail: `Failed to update recipe: ${(err as any).message}` }, 400)
  }
})

router.delete('/:id', async (c) => {
  const id = c.req.param('id')
  const currentUser = await getCurrentUser(c)
  if (currentUser.role !== 'cms_editor') {
    return c.json({ detail: 'Forbidden: Requires cms_editor role.' }, 403)
  }

  const serviceClient = getServiceClient()
  try {
    const { error } = await serviceClient.from('recipes').delete().eq('id', id)
    if (error) {
      return c.json({ detail: `Failed to delete recipe: ${error.message}` }, 400)
    }
    c.status(204)
    return c.body(null)
  } catch (err) {
    return c.json({ detail: `Failed to delete recipe: ${(err as any).message}` }, 400)
  }
})

export default router
